File: test/Zametek.Maths.Graphs.Primitives.Tests/Entities/ActivityTests.cs
using Shouldly;
using System;
using System.Linq;
using Xunit;

namespace Zametek.Maths.Graphs.Tests
{
    public class ActivityTests
    {
        [Fact]
        public void Activity_GivenCtor_ThenPropertiesSetAndCollectionsEmpty()
        {
            var activity = new Activity<int, int, int>(1, 5);

            activity.Id.ShouldBe(1);
            activity.Duration.ShouldBe(5);
            activity.TargetWorkStreams.ShouldBeEmpty();
            activity.TargetResources.ShouldBeEmpty();
            activity.AllocatedToResources.ShouldBeEmpty();
            activity.CanBeRemoved.ShouldBeFalse();
        }

        [Fact]
        public void Activity_GivenEarliestFinishTime_WhenNoEarliestStartTime_ThenNull()
        {
            var activity = new Activity<int, int, int>(1, 5);

            activity.EarliestFinishTime.ShouldBeNull();
        }

        [Fact]
        public void Activity_GivenEarliestFinishTime_WhenEarliestStartTimeSet_ThenStartPlusDuration()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 3,
            };

            activity.EarliestFinishTime.ShouldBe(8);
        }

        [Fact]
        public void Activity_GivenLatestStartTime_WhenLatestFinishTimeSet_ThenFinishMinusDuration()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                LatestFinishTime = 12,
            };

            activity.LatestStartTime.ShouldBe(7);
        }

        [Fact]
        public void Activity_GivenTotalSlack_WhenBothFinishTimesAvailable_ThenLatestMinusEarliest()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,   // EF = 5
                LatestFinishTime = 8,
            };

            activity.TotalSlack.ShouldBe(3);
        }

        [Fact]
        public void Activity_GivenTotalSlack_WhenTimesMissing_ThenNull()
        {
            var activity = new Activity<int, int, int>(1, 5);

            activity.TotalSlack.ShouldBeNull();
        }

        [Fact]
        public void Activity_GivenInterferingSlack_ThenTotalSlackMinusFreeSlack()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,   // EF = 5
                LatestFinishTime = 8,    // total slack = 3
                FreeSlack = 1,
            };

            activity.InterferingSlack.ShouldBe(2);
        }

        [Fact]
        public void Activity_GivenInterferingSlack_WhenFreeSlackMissing_ThenNull()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,
                LatestFinishTime = 8,
            };

            activity.InterferingSlack.ShouldBeNull();
        }

        [Fact]
        public void Activity_GivenIsCritical_WhenZeroTotalSlack_ThenTrue()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,
                LatestFinishTime = 5,
            };

            activity.IsCritical.ShouldBeTrue();
        }

        [Fact]
        public void Activity_GivenIsCritical_WhenNegativeTotalSlack_ThenTrue()
        {
            // Over-constrained: latest finish before earliest finish.
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,
                LatestFinishTime = 3,
            };

            activity.TotalSlack.ShouldBe(-2);
            activity.IsCritical.ShouldBeTrue();
        }

        [Fact]
        public void Activity_GivenIsCritical_WhenPositiveTotalSlack_ThenFalse()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                EarliestStartTime = 0,
                LatestFinishTime = 8,
            };

            activity.IsCritical.ShouldBeFalse();
        }

        [Fact]
        public void Activity_GivenIsCritical_WhenNoTimes_ThenFalse()
        {
            var activity = new Activity<int, int, int>(1, 5);

            activity.IsCritical.ShouldBeFalse();
        }

        [Fact]
        public void Activity_GivenIsDummy_WhenZeroDuration_ThenTrue()
        {
            new Activity<int, int, int>(1, 0).IsDummy.ShouldBeTrue();
            new Activity<int, int, int>(1, 1).IsDummy.ShouldBeFalse();
        }

        [Fact]
        public void Activity_GivenSetAsRemovableAndReadOnly_ThenCanBeRemovedToggles()
        {
            var activity = new Activity<int, int, int>(1, 5);

            activity.SetAsRemovable();
            activity.CanBeRemoved.ShouldBeTrue();

            activity.SetAsReadOnly();
            activity.CanBeRemoved.ShouldBeFalse();
        }

        [Fact]
        public void Activity_GivenCloneObject_ThenAllPropertiesPreserved()
        {
            var activity = new Activity<int, int, int>(1, 5)
            {
                Name = "name",
                Notes = "notes",
                TargetResourceOperator = LogicalOperator.OR,
                HasNoCost = true,
                HasNoBilling = true,
                HasNoEffort = true,
                EarliestStartTime = 2,
                LatestFinishTime = 10,
                FreeSlack = 1,
                MinimumFreeSlack = 1,
                MinimumEarliestStartTime = 2,
                MaximumLatestFinishTime = 11,
            };
            activity.TargetWorkStreams.Add(21);
            activity.TargetResources.Add(31);
            activity.AllocatedToResources.Add(41);
            activity.SetAsRemovable();

            var clone = (Activity<int, int, int>)activity.CloneObject();

            clone.Id.ShouldBe(1);
            clone.Duration.ShouldBe(5);
            clone.Name.ShouldBe("name");
            clone.Notes.ShouldBe("notes");
            clone.TargetResourceOperator.ShouldBe(LogicalOperator.OR);
            clone.HasNoCost.ShouldBeTrue();
            clone.HasNoBilling.ShouldBeTrue();
            clone.HasNoEffort.ShouldBeTrue();
            clone.EarliestStartTime.ShouldBe(2);
            clone.LatestFinishTime.ShouldBe(10);
            clone.FreeSlack.ShouldBe(1);
            clone.MinimumFreeSlack.ShouldBe(1);
            clone.MinimumEarliestStartTime.ShouldBe(2);
            clone.MaximumLatestFinishTime.ShouldBe(11);
            clone.TargetWorkStreams.ShouldBe([21]);
            clone.TargetResources.ShouldBe([31]);
            clone.AllocatedToResources.ShouldBe([41]);
            clone.CanBeRemoved.ShouldBeTrue();
        }

        [Fact]
        public void Activity_GivenCloneObject_ThenCollectionsAreIndependentCopies()
        {
            var activity = new Activity<int, int, int>(1, 5);
            activity.TargetResources.Add(31);

            var clone = (Activity<int, int, int>)activity.CloneObject();
            clone.TargetResources.Add(32);

            activity.TargetResources.Count.ShouldBe(1);
            clone.TargetResources.Count.ShouldBe(2);
        }
    }
}

File: rust/primitives/tests/activity_tests.rs
//! Ports of `Zametek.Maths.Graphs.Primitives.Tests/Entities/ActivityTests.cs`.

use indexmap::IndexSet;
use zametek_maths_graphs_primitives::{Activity, LogicalOperator};

type Act = Activity<i32, i32, i32>;

#[test]
fn activity_given_ctor_then_properties_set_and_collections_empty() {
    let activity = Act::new(1, 5);
    assert_eq!(activity.id(), 1);
    assert_eq!(activity.duration, 5);
    assert!(activity.target_work_streams.is_empty());
    assert!(activity.target_resources.is_empty());
    assert!(activity.allocated_to_resources.is_empty());
    assert!(!activity.can_be_removed());
}

#[test]
fn activity_given_earliest_finish_time_when_no_earliest_start_time_then_none() {
    let activity = Act::new(1, 5);
    assert_eq!(activity.earliest_finish_time(), None);
}

#[test]
fn activity_given_earliest_finish_time_when_earliest_start_time_set_then_start_plus_duration() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(3);
    assert_eq!(activity.earliest_finish_time(), Some(8));
}

#[test]
fn activity_given_latest_start_time_when_latest_finish_time_set_then_finish_minus_duration() {
    let mut activity = Act::new(1, 5);
    activity.latest_finish_time = Some(12);
    assert_eq!(activity.latest_start_time(), Some(7));
}

#[test]
fn activity_given_total_slack_when_both_finish_times_available_then_latest_minus_earliest() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0); // EF = 5
    activity.latest_finish_time = Some(8);
    assert_eq!(activity.total_slack(), Some(3));
}

#[test]
fn activity_given_total_slack_when_times_missing_then_none() {
    let activity = Act::new(1, 5);
    assert_eq!(activity.total_slack(), None);
}

#[test]
fn activity_given_interfering_slack_then_total_slack_minus_free_slack() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0); // EF = 5
    activity.latest_finish_time = Some(8); // total slack = 3
    activity.free_slack = Some(1);
    assert_eq!(activity.interfering_slack(), Some(2));
}

#[test]
fn activity_given_interfering_slack_when_free_slack_missing_then_none() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0);
    activity.latest_finish_time = Some(8);
    assert_eq!(activity.interfering_slack(), None);
}

#[test]
fn activity_given_is_critical_when_zero_total_slack_then_true() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0);
    activity.latest_finish_time = Some(5);
    assert!(activity.is_critical());
}

#[test]
fn activity_given_is_critical_when_negative_total_slack_then_true() {
    // Over-constrained: latest finish before earliest finish.
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0);
    activity.latest_finish_time = Some(3);
    assert_eq!(activity.total_slack(), Some(-2));
    assert!(activity.is_critical());
}

#[test]
fn activity_given_is_critical_when_positive_total_slack_then_false() {
    let mut activity = Act::new(1, 5);
    activity.earliest_start_time = Some(0);
    activity.latest_finish_time = Some(8);
    assert!(!activity.is_critical());
}

#[test]
fn activity_given_is_critical_when_no_times_then_false() {
    let activity = Act::new(1, 5);
    assert!(!activity.is_critical());
}

#[test]
fn activity_given_is_dummy_when_zero_duration_then_true() {
    assert!(Act::new(1, 0).is_dummy());
    assert!(!Act::new(1, 1).is_dummy());
}

#[test]
fn activity_given_removable_ctor_then_can_be_removed() {
    assert!(Act::new_removable(1, 5, true).can_be_removed());
    assert!(!Act::new_removable(1, 5, false).can_be_removed());
}

#[test]
fn activity_given_set_as_removable_and_read_only_then_can_be_removed_toggles() {
    let mut activity = Act::new(1, 5);
    activity.set_as_removable();
    assert!(activity.can_be_removed());
    activity.set_as_read_only();
    assert!(!activity.can_be_removed());
}

#[test]
fn activity_given_clone_then_all_properties_preserved() {
    let mut activity = Act::new(1, 5);
    activity.name = Some("name".to_string());
    activity.notes = Some("notes".to_string());
    activity.target_resource_operator = LogicalOperator::Or;
    activity.has_no_cost = true;
    activity.has_no_billing = true;
    activity.has_no_effort = true;
    activity.earliest_start_time = Some(2);
    activity.latest_finish_time = Some(10);
    activity.free_slack = Some(1);
    activity.minimum_free_slack = Some(1);
    activity.minimum_earliest_start_time = Some(2);
    activity.maximum_latest_finish_time = Some(11);
    activity.target_work_streams.insert(21);
    activity.target_resources.insert(31);
    activity.allocated_to_resources.insert(41);
    activity.set_as_removable();

    let clone = activity.clone();

    assert_eq!(clone.id(), 1);
    assert_eq!(clone.duration, 5);
    assert_eq!(clone.name.as_deref(), Some("name"));
    assert_eq!(clone.notes.as_deref(), Some("notes"));
    assert_eq!(clone.target_resource_operator, LogicalOperator::Or);
    assert!(clone.has_no_cost);
    assert!(clone.has_no_billing);
    assert!(clone.has_no_effort);
    assert_eq!(clone.earliest_start_time, Some(2));
    assert_eq!(clone.latest_finish_time, Some(10));
    assert_eq!(clone.free_slack, Some(1));
    assert_eq!(clone.minimum_free_slack, Some(1));
    assert_eq!(clone.minimum_earliest_start_time, Some(2));
    assert_eq!(clone.maximum_latest_finish_time, Some(11));
    assert_eq!(clone.target_work_streams, IndexSet::from([21]));
    assert_eq!(clone.target_resources, IndexSet::from([31]));
    assert_eq!(clone.allocated_to_resources, IndexSet::from([41]));
    assert!(clone.can_be_removed());
}

#[test]
fn activity_given_clone_then_collections_are_independent_copies() {
    let mut activity = Act::new(1, 5);
    activity.target_resources.insert(31);

    let mut clone = activity.clone();
    clone.target_resources.insert(32);

    assert_eq!(activity.target_resources.len(), 1);
    assert_eq!(clone.target_resources.len(), 2);
}
